File: GeeksforGeeks/Must Do Coding Questions for Product Based Companies/Arrays/Next Permutation.py
#https://practice.geeksforgeeks.org/problems/next-permutation5226/1

#User function Template for python3

class Solution:
    def nextPermutation(self, N, arr):
        # code here
        if (arr==sorted(arr,reverse=True)):
            arr=sorted(arr)
            return arr
        else:    
            for i in range(N-1,0,-1):
                if(arr[i-1]<arr[i]):
                    arr=arr[0:i]+sorted(arr[i:N])
                    for j in range(i,N):
                        if(arr[i-1]<arr[j]):
                            temp=arr[j]
                            arr[j]=arr[i-1]
                            arr[i-1]=temp
                            return arr
        return arr
#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        N = int(input())
        arr = input().split()
        for i in range(N):
            arr[i] = int(arr[i])
        
        ob = Solution()
        ans = ob.nextPermutation(N, arr)
        for i in range(N):
            print(ans[i],end=" ")
        print()
# } Driver Code Ends

File: QuesByBabbarFINAL450/Next Permutation.cpp
//https://leetcode.com/problems/next-permutation/

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
       int n=nums.size();
        for(int i=n-1;i>=1;i--)
        {
            if(nums[i]>nums[i-1])
            {
                int temp=nums[i-1];
                for(int j=n-1;j>=i;j--)
                {
                    if(nums[j]>temp)
                    {
                        swap(nums[i-1],nums[j]);
                        sort(nums.begin()+i,nums.end());
                        return ;
                    }
                }
            }
        }
        sort(nums.begin(),nums.end());
        return ;
    }
};
